utils/typename-derive/src/lib.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro::{self, TokenStream};
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput};
+
+#[proc_macro_derive(TypeName)]
+pub fn derive(input: TokenStream) -> TokenStream {
+    let DeriveInput {
+        ident,
+        attrs: _attrs,
+        vis: _vis,
+        generics,
+        data: _data,
+    } = parse_macro_input!(input);
+    let output = if generics.params.is_empty() {
+        let type_name = ident.to_string();
+        quote! {
+            impl TypeName for #ident {
+                fn typename_str() -> std::borrow::Cow<'static, str> {
+                    #type_name.into()
+                }
+            }
+        }
+    } else {
+        let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+        let type_name = ident.to_string();
+        quote! {
+            impl #impl_generics TypeName for #ident #ty_generics #where_clause {
+                fn typename_str() -> std::borrow::Cow<'static, str> {
+                    std::borrow::Cow::Owned(#type_name.to_owned() + "<" + T::typename_str().as_ref() + ">")
```

Comment:
Does this work? I don't see any `T` in scope for `T::typename_str()` to work. Also I'd expect there to be one cal to `T::typename_str()` per generic parameter, not just one for all params.

Reply:
You're right. I'll look into it.

Reply:
Just checked: is only works if a structure has only one generic parameter named `T`.

Reply:
I guess it is possible to get `TokenStream` from `TypeGenerics` (`ty_generics`) and get a list of type names from this stream, but it won't be easy.

Reply:
I'm doing that actually as we speak. But it needs some work.

Reply:
Issue is resolved.